featurebyte/feast/patch.py

```diff
@@ -202,3 +203,104 @@ def with_projection(
+    Tha main difference between this and the original Feast implementation is that this function converts the
+    arrow table to a pandas dataframe first and then call `transform` method of the feature view to transform
+    the features. DataFrame with alias column names support is used to improve the runtime & memory performance
+    of the function without copying the columns & dropping after transformation. This significantly reduces the
+    runtime & memory usage of the function.
+
+    Parameters
+    ----------
+    feature_view: OnDemandFeatureView
+        OnDemandFeatureView object
+    pa_table: pyarrow.Table
+        Arrow table to transform
+    full_feature_names: bool
+        A boolean that provides the option to add the feature view prefixes to the feature names,
+
+    Returns
+    -------
+    pyarrow.Table
+        PyArrow table with transformed features
+
+    Raises
+    ------
+    TypeError
+        transform_arrow only accepts pyarrow.Table
+    """
+    if not isinstance(pa_table, pyarrow.Table):
+        raise TypeError("transform_arrow only accepts pyarrow.Table")
+
+    # Transform the arrow to a pandas dataframe first
+    input_df = pa_table.to_pandas()
+
+    # Original implementation assigns a new column on each iteration, this implementation uses a wrapper
+    # class to improve the runtime & memory performance.
+    df_with_features = DataFrameWrapper(input_df)
+    for source_fv_projection in feature_view.source_feature_view_projections.values():
+        for feature in source_fv_projection.features:
+            full_feature_ref = f"{source_fv_projection.name}__{feature.name}"
+            if full_feature_ref in df_with_features.keys():
+                # Make sure the partial feature name is always present
+                df_with_features.add_column_alias(full_feature_ref, feature.name)
+            elif feature.name in df_with_features.keys():
+                # Make sure the full feature name is always present
+                df_with_features.add_column_alias(feature.name, full_feature_ref)
+
+    # Compute transformed values and apply to each result row
+    df_with_transformed_features = feature_view.feature_transformation.transform(df_with_features)
+    assert isinstance(df_with_transformed_features, pd.DataFrame)
+
+    # Work out whether the correct columns names are used.
+    rename_columns: Dict[str, str] = {}
+    for feature in feature_view.features:
+        short_name = feature.name
+        long_name = feature_view._get_projected_feature_name(feature.name)
+        if short_name in df_with_transformed_features.columns and full_feature_names:
+            rename_columns[short_name] = long_name
+        elif not full_feature_names:
+            # Long name must be in dataframe.
+            rename_columns[long_name] = short_name
+
+    df_with_features = pd.DataFrame(df_with_transformed_features).rename(columns=rename_columns)
+    return pyarrow.Table.from_pandas(df_with_features)
```

Comment:
This is similar to https://github.com/featurebyte/featurebyte/blob/2fdd377e7c3e8e71ab47fc9516fd131ba307d089/featurebyte/feast/patch.py#L137-L188. The main difference is that we transform the `pyarrow.Table` to `pd.DataFrame` at the beginning & convert it back before returning the result.